File: java/org/rocksdb/Filter.java
// Copyright (c) 2014, Facebook, Inc.  All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.

package org.rocksdb;

/**
 * Filters are stored in rocksdb and are consulted automatically
 * by rocksdb to decide whether or not to read some
 * information from disk. In many cases, a filter can cut down the
 * number of disk seeks form a handful to a single disk seek per
 * DB::Get() call.
 *
 * This function a new filter policy that uses a bloom filter
 * with approximately the specified number of bits per key.
 * A good value for bitsPerKey is 10, which yields a filter
 * with ~ 1% false positive rate. 
 */
public class Filter {
  private long nativeHandle_;
    
  public Filter(int bitsPerKey) {
    newFilter(bitsPerKey);
  }
  
  public long getNativeHandle() {
    return nativeHandle_;
  }
  
  /**
   * Deletes underlying C++ filter pointer.
   */
  public synchronized void dispose() {
    if(nativeHandle_ != 0) {
      dispose0(nativeHandle_);
    }
  }

  @Override protected void finalize() {
    dispose();
  }

  private boolean isInitialized() {
    return (nativeHandle_ != 0);
  }
  
  private native void newFilter(int bitsPerKey);
  private native void dispose0(long handle);
}

File: java/rocksjni/filter.cc
// Copyright (c) 2014, Facebook, Inc.  All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree. An additional grant
// of patent rights can be found in the PATENTS file in the same directory.
//
// This file implements the "bridge" between Java and C++ for
// rocksdb::FilterPolicy.

#include <stdio.h>
#include <stdlib.h>
#include <jni.h>
#include <string>

#include "include/org_rocksdb_Filter.h"
#include "rocksjni/portal.h"
#include "rocksdb/filter_policy.h"

/*
 * Class:     org_rocksdb_Filter
 * Method:    newFilter
 * Signature: (I)V
 */
void Java_org_rocksdb_Filter_newFilter(
    JNIEnv* env, jobject jobj, jint bits_per_key) {
  const rocksdb::FilterPolicy* fp = rocksdb::NewBloomFilterPolicy(bits_per_key);
  rocksdb::FilterJni::setHandle(env, jobj, fp);
}

/*
 * Class:     org_rocksdb_Filter
 * Method:    dispose0
 * Signature: (J)V
 */
void Java_org_rocksdb_Filter_dispose0(
    JNIEnv* env, jobject jobj, jlong handle) {
  auto fp = reinterpret_cast<rocksdb::FilterPolicy*>(handle);
  delete fp;

  rocksdb::FilterJni::setHandle(env, jobj, nullptr);
}
